added step_discretize_tree which performs discretisation in a supervised way using xgboost

konradsemsch opened the pull request:
The most recent version of this PR: tidymodels/recipes#381

How should be go about tests and tunable methods? I would be happy to continue here but I would need a bit of guidance on what should be done.

topepo commented:
If you've never made unit tests before, give [this a read](http://r-pkgs.had.co.nz/tests.html). 

There are a lot of examples in the `tests` directory. Basically, use small data sets where you can figure out the correct results and test that you get those results. `covr::package_coverage()` is helpful to figure out how well tested the code is. 

From the check results, you have some `:::` in the package that need to be `::`. 

Also, there are false positives for global variables, such as `xgb_binning: no visible binding for global variable ‘Node’`. You can remove those by adding `utils::globalVariables(values)` where `values` is a character vector of symbols that it is complaining about. Make sure that these are false positives though. 

For `tunable`, you would create an S3 method for which parameters that you want to tune, their `dials` counterparts, and (optionally) any specific parameter ranges that you think should be used by default. Lots of examples can be found [here](https://github.com/tidymodels/tune/blob/fa4dc535e89ad8d039df5ba5717e718485d877f4/R/tunable.R#L114). 

I'll probably fork you repo and kick it around a bit over the break. I might have some suggestions then.

topepo commented:
See [the changes](https://github.com/topepo/embed-1/commit/795d7f18f961eaf478ca338f20e2c97b9955d690) that are in my fork of your fork. I can do a PR to your repo if you want them. 

I fixed a few bugs (regression wasn't working, all of the breaks were not saved, etc). I also think that we need two more arguments (one for the number of unique values to check for and another for the size of the validation set).  

CRAN wants me to have another release by 2020/01/15 so let me know if you want to shoot for that.

topepo commented:
So I think that we need some unit tests. In my PR I also added some commnents with a `TODO` tag:

```
$ grep -C 2 TODO *.R
discretize_tree.R-           outcome = NULL,
discretize_tree.R-           learn_rate = 0.3,
discretize_tree.R:           num_breaks = 10, # TODO maybe default this to a smaller value for smaller data sets. 
discretize_tree.R-           tree_depth = 1,
discretize_tree.R-           rules = NULL,
--
--
discretize_tree.R-           tree_depth = 1,
discretize_tree.R-           rules = NULL,
discretize_tree.R:           # TODO I don't think that we need to rename since we are replacing with a single column
discretize_tree.R-           prefix = "bin_tree",
discretize_tree.R-           skip = FALSE,
--
--
discretize_tree.R-  # ----------------------------------------------------------------------------
discretize_tree.R-  
discretize_tree.R:  # TODO prop should probably be a parameter
discretize_tree.R-  # What to do if it is 1 (or bound it so that test has at least 2 data points)?
discretize_tree.R-  split <- rsample::initial_split(df, prop = 0.8, strata = outcome)
--
--
discretize_tree.R-  
discretize_tree.R-  col_names <- col_names[col_names != y_name]
discretize_tree.R:  # TODO test to see if length(col_names) >= 1 here
discretize_tree.R-  
--
discretize_tree.R-  
discretize_tree.R:  # TODO Have a check for the minimum number of unique data points in the column
discretize_tree.R-  # in order to run the step. Otherwise, take it out of col_names. I think that 
discretize_tree.R-  # num_unique = 20 is probably a good default. 
--
--
discretize_tree.R-    var <- names(vars)[[i]]
discretize_tree.R-    
discretize_tree.R:    # TODO not sure that you need to make a copy
discretize_tree.R-    
discretize_tree.R-    binned_data <- new_data
--
--
discretize_tree.R-
discretize_tree.R-
discretize_tree.R:# TODO a tidy() method that returns a tibble with columns for terms (the column
discretize_tree.R-# names) and `values` (the break values))
discretize_tree.R-
```

konradsemsch commented:
No worries, I'm already working on that. I will push a new version by the end of the weekend.

konradsemsch commented:
I worked on adding a number of different tests for different scenarios (I think 2 are not yet working, but will wait for your answer), and while working on them I had to make a number of changes/ improvements in the main functions. I added comments/ TODOs for us to go over them to clarify and finalise. 

The biggest things to finish/ discuss are:

- `tidy` method - getting a weird error that I don't understand. Was trying to debug but with no luck. The details of the error and comments are at the end of the test file
- `tunable` method - please take a look if that's correct. I also made some comments there
- **main file:**
1. Exposing `nrounds`/ `early_stopping_rounds` to the user to potentially fine-tune? I think we might keep it as is actually for now and eventually change that in the next pass?
2. My error handling implementation when XgBoost returns a constant model - what I implemented worked for the examples I tested but I don't think it's a very neat way of doing it
3. I didn't get your comment regarding copying data (`bake` method)

topepo commented:
Just an fyi.. I need to do an `embed` release in the next two weeks. I'm going to work in my fork of your fork and make some updates. I'll do a PR back into yours in the next few days.

konradsemsch commented:
Cool, looking forward to bring this to an end!

topepo commented:
> Cool, looking forward to bring this to an end!

Out it out of its misery? 😆 

Here's question: the underlying binning code converts data [to integer](https://github.com/konradsemsch/embed-1/blob/c6aa2bb8208406ce4e3468e150d0a3897b1f6baa/R/discretize_tree.R#L202). Is this required by xgboost?

topepo commented:
If they need to be integer then we might want to rescale them to a set range (maybe 0 to 1000?) so that data that are on `[0, 1]` aren't squashed. 

Also, `step_integer()` is really for converting character data (or factors) into integers. Using it in numeric data produces unexpected results (which is how I got here).

All-in-all I'd rather leave them as-is unless xgboost really needs integers.

topepo commented:
Kicking things around a bit, I'm seeing it overfit a fair amount when the outcome is a number: 

``` r
# devtools::install_github("topepo/embed-1")
library(embed)
#> Loading required package: recipes
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(rpart)
library(ggplot2)
theme_set(theme_bw())
```

``` r

library(dplyr)

sim_data_2class <- function(n = 300) {
  xf <- function(x) {
    case_when(
      x < 0.33 ~ "a",
      x < 0.66 & x >= 0.33 ~ "b",
      TRUE ~ "a"
    )
  }
  
  dat <- tibble(x = runif(n), z = runif(n)) %>% 
    mutate(class = purrr::map_chr(x, xf),
           class = factor(class, levels = letters[1:2]))
  dat
}

sim_data_reg <- function(n = 300) {
  xf <- function(x) {
    case_when(
      x < 0.33 ~ 1,
      x < 0.66 & x >= 0.33 ~ 3,
      TRUE ~ 5
    )
  }
  
  dat <- tibble(x = runif(n), z = runif(n)) %>% 
    mutate(y = purrr::map_dbl(x, xf),
           y = y + rnorm(n, sd = .1))
  dat
}
```

``` r

set.seed(38384)
dat_cls <- sim_data_2class(n = 300) # bins at 1/3 and 2/3

xgb_cls <-
  recipe(class ~ ., data = dat_cls) %>%
  step_discretize_tree(all_predictors(), outcome = "class") %>% 
  prep()

ggplot(dat_cls, aes(x = x)) + 
  geom_histogram() + 
  facet_wrap(~ class, ncol = 1) + 
  geom_vline(xintercept = xgb_cls$steps[[1]]$rules$x, lty = 12)
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

![](https://i.imgur.com/NIDdAMT.png)

``` r

set.seed(38384)
dat_reg <- sim_data_reg(n = 300) # bins at 1/3 and 2/3

xgb_reg <-
  recipe(y ~ ., data = dat_reg) %>%
  step_discretize_tree(all_predictors(), outcome = "y") %>% 
  prep()

ggplot(dat_reg, aes(x = x, y = y)) + 
  geom_point(alpha = .3) + 
  geom_vline(xintercept = xgb_reg$steps[[1]]$rules$x, lty = 12)
```

![](https://i.imgur.com/SdXhNs0.png)

``` r

rpart::rpart(y ~ x, data = dat_reg)
#> n= 300 
#> 
#> node), split, n, deviance, yval
#>       * denotes terminal node
#> 
#> 1) root 300 825.5522000 3.0131400  
#>   2) x< 0.3296669 102   0.8233053 0.9966416 *
#>   3) x>=0.3296669 198 196.3061000 4.0519420  
#>     6) x< 0.659105 93   0.9744730 3.0001350 *
#>     7) x>=0.659105 105   1.3185650 4.9835430 *
```

<sup>Created on 2020-05-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I know that it is not as cool, but `rpart` might be a better engine (and it works for 3+ classes).

konradsemsch commented:
> Here's question: the underlying binning code converts data to integer. Is this required by xgboost?

I couldn't remember all decisions that I made back then since some time has passed, but I quick check confirms that: [link]. (https://www.rdocumentation.org/packages/xgboost/versions/1.0.0.2/topics/xgb.DMatrix). So yes, we need that integer conversion when working with XgBoost.

> If they need to be integer then we might want to rescale them to a set range (maybe 0 to 1000?) so that data that are on [0, 1] aren't squashed. Also, step_integer() is really for converting character data (or factors) into integers. Using it in numeric data produces unexpected results (which is how I got here).

Hmm, I think I would then comvert to integers only character/ factor data and leave numeric data untouched in it's original form. XgBoost should handle that pretty well natively and perhaps we could leave those decisions to the modeller (prior applying this step)?

> Kicking things around a bit, I'm seeing it overfit a fair amount when the outcome is a number. I know that it is not as cool, but rpart might be a better engine (and it works for 3+ classes).

That was a little bit my fear as well when I was testing it on rather small datasets... The results were much better when working on bigger ones.  Not sure if `rpart` will offer good enough performance for those (also no early stopping to fight overfitting). My thinking right now:

- we can also cover multiclass with `xgboost` using "multi:softprob" objective
- for smaller datasets we could additionally fight overfitting by controlling `min_child_weight ` (potentially exposing it to the user)?
- we could consider another resample strategy to perform early stopping for smaller datasets to find the best results through averaging them
- we could also implement two backends: rpart vs. xgboost, and allow the user to choose one fo them depending on the nature of the dataset

What do you think?

topepo commented:
Looking at the `xgb.DMatrix()` docs (and testing) the data do not have to be integers so I would leave them as-is. 

Honestly I think that `rpart` is a better choice but we can enable both. 

One other question: if there is a failure or some other issue preventing the model from determining breaks, are you okay with returning the predictor in its original form and issuing a warning? That's a lot smoother than erroring out.

konradsemsch commented:
> Looking at the xgb.DMatrix() docs (and testing) the data do not have to be integers so I would leave them as-is.

Have you tested it? XgBoost matrix data argument says: "a matrix object (**either numeric or integer)**, a dgCMatrix object, or a character string representing a filename."

> One other question: if there is a failure or some other issue preventing the model from determining breaks, are you okay with returning the predictor in its original form and issuing a warning? That's a lot smoother than erroring out.

Sure - you have much more experience building these solutions, so let's follow this approach as you suggest.

topepo commented:
The raw data fall under "numeric" so we're good. The fork that I have has simulated tests that give the right results. Please give it a look over just to make sure though.

topepo commented:
Take a look at [this fork](https://github.com/topepo/embed-1). It passes the checks (locally) but there are some failures on GH actions till. It seems to be related to reproducibility of xgb models. I changed the step suffix to `_xgb`. 

I also added a CART version too.

konradsemsch commented:
Nice job! I would have two comments for now:

- I think it's worth adding the multiclass objective for XgBoost as well to make sure both methods can be applied to the same range of scenarios

- don't you think that `three_depth` of 10 for cart is a bit excessive as default? The corresponding value is 1 in the XgBoost implementation. I think that with 10 it might often lead to overfitting.

topepo commented:
👍 to multiclass xgb

For CART there's not much to worry about in terms of overfitting. If you try really hard to can make that happen but generally the pruning process will settle on a fairly parsimonious fit. 

For example, with the simulated data: 

```r
> rpart(y ~ x, data = sim_data_reg(10000), maxdepth = 20)
n= 10000 

node), split, n, deviance, yval
      * denotes terminal node

1) root 10000 27013.41000 3.020806  
  2) x< 0.659897 6579  6614.37700 1.991735  
    4) x< 0.3302054 3317    32.28917 1.002334 *
    5) x>=0.3302054 3262    33.22668 2.997818 *
  3) x>=0.659897 3421    33.39948 4.999836 *
```

Disabling pruning will get you there though: 

```r
> length(rpart(y ~ x, data = sim_data_reg(10000), maxdepth = 20, cp = 0)$splits[,"index"])
[1] 446
```

konradsemsch commented:
Ok. Will you add the multiclass implementation in your fork?

topepo commented:
Can you make the multi-class changes? I can lend a hand if you need it. 

I've pretty much finalized the fork. There is still an error for R 3.5 that I need to look into; it should be skipping some tests and it isn't. 

Let me know if there is anything substantial that you'd like to change.

konradsemsch commented:
Sure, I have time tomorrow to take care of it. Will you merge?

topepo commented:
Yes. Update your fork with my fork, make the changes, then that should carry over to this PR (then I'll review).

konradsemsch commented:
I built on top of your fork and here's a summary of my commit:

> added multi-class for xgb

> added additional tests for xgb multiclass

> exposed `min_n` as a parameter to the user with the previous default

> corrected some failing tests unrelated to this implementation

I also skimmed through the tunable methods and I wanted to ask you to take a look at that (I think it was previously failing for me). Are they implemented correctly? I can't see any tunable method for your CART implementation.

topepo commented:
For the `tunable()` method: 

 >  do I assume correctly that if no new range is given, dials defaults are assumed?

Yes. 

>  is it possible to define a set of integers instead of range?

No but using `c(5, 30)` effectively makes the set to be `5:30`. 

> I can't see any tunable method for your CART implementation.

Thanks. I forgot that. I'll add it. 

> I think it was previously failing for me

Can you send an example? 

I'm going to look into this more but any tuning that I've do on the learning rate leads to trees with no splits (code below). The xgboost default is 0.3. The default range for the `dials` method has a high end of 0.1. Perhaps we should use a custom range. For the Ames example, values below 0.01 in my testing failed. There are some articles about this too. 

I'm doing some testing now and, for main arguments, we add an entry for `tunable()`. Can you add one for `prop()` (but see next comment)? 

For `prop()`, can we 

 * rename is to `sample_val()` and reverse the definition? 

 * make a `dials` object with a range of 0.05 to 0.20 (or something similar)
 
 * add an entry to `tunable()`

Example: 

``` r
library(tidymodels)
#> ── Attaching packages ──────────────────────────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6      ✓ recipes   0.1.12
#> ✓ dials     0.0.6      ✓ rsample   0.0.6 
#> ✓ dplyr     0.8.5      ✓ tibble    3.0.1 
#> ✓ ggplot2   3.3.0      ✓ tune      0.1.0 
#> ✓ infer     0.5.1      ✓ workflows 0.1.1 
#> ✓ parsnip   0.1.1      ✓ yardstick 0.0.6 
#> ✓ purrr     0.3.4
#> Warning: package 'broom' was built under R version 3.6.2
#> Warning: package 'dials' was built under R version 3.6.2
#> Warning: package 'scales' was built under R version 3.6.2
#> Warning: package 'recipes' was built under R version 3.6.2
#> Warning: package 'rsample' was built under R version 3.6.2
#> Warning: package 'tune' was built under R version 3.6.2
#> ── Conflicts ─────────────────────────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard()  masks scales::discard()
#> x dplyr::filter()   masks stats::filter()
#> x dplyr::lag()      masks stats::lag()
#> x ggplot2::margin() masks dials::margin()
#> x recipes::step()   masks stats::step()
library(AmesHousing)
library(embed)
```

``` r

ames <-
  make_ames() %>%
  # Remove quality-related predictors
  dplyr::select(-matches("Qu"))

set.seed(4595)
data_split <- initial_split(ames, strata = "Sale_Price")
ames_train <- training(data_split)
ames_test  <- testing(data_split)
rs <- vfold_cv(ames_train)
```

``` r

ames_rec <-
  recipe(Sale_Price ~ Longitude + Latitude,
         data = ames_train) %>%
  step_log(Sale_Price, base = 10) %>%
  step_discretize_xgb(
    Longitude,
    outcome = "Sale_Price",
    min_n = tune("lon min_n"),
    num_breaks = tune("lon breaks"),
    tree_depth = tune("lon depth"),
    learn_rate = .01
  ) %>%
  step_discretize_xgb(
    Latitude,
    outcome = "Sale_Price",
    min_n = tune("lat min_n"),
    num_breaks = tune("lat breaks"),
    tree_depth = tune("lat depth"),
    learn_rate = .01
  ) %>%
  step_zv(all_predictors())

mod <- linear_reg() %>% set_engine("lm")

set.seed(8199)
res <-
  mod %>% tune_grid(
    ames_rec,
    resamples = rs,
    grid = 10,
    control = control_grid(verbose = TRUE)
  )
#> i Fold01: recipe  1/10
#> ✓ Fold01: recipe  1/10
#> i Fold01: recipe  1/10, model 1/1
#> ✓ Fold01: recipe  1/10, model 1/1
#> i Fold01: recipe  1/10, model 1/1 (predictions)
#> i Fold01: recipe  2/10
#> ✓ Fold01: recipe  2/10
#> i Fold01: recipe  2/10, model 1/1
#> ✓ Fold01: recipe  2/10, model 1/1
#> i Fold01: recipe  2/10, model 1/1 (predictions)
#> i Fold01: recipe  3/10
#> ✓ Fold01: recipe  3/10
#> i Fold01: recipe  3/10, model 1/1
#> ✓ Fold01: recipe  3/10, model 1/1
#> i Fold01: recipe  3/10, model 1/1 (predictions)
#> i Fold01: recipe  4/10
#> ✓ Fold01: recipe  4/10
#> i Fold01: recipe  4/10, model 1/1
#> ✓ Fold01: recipe  4/10, model 1/1
#> i Fold01: recipe  4/10, model 1/1 (predictions)
#> i Fold01: recipe  5/10
#> ✓ Fold01: recipe  5/10
#> i Fold01: recipe  5/10, model 1/1
#> ✓ Fold01: recipe  5/10, model 1/1
#> i Fold01: recipe  5/10, model 1/1 (predictions)
#> i Fold01: recipe  6/10
#> ✓ Fold01: recipe  6/10
#> i Fold01: recipe  6/10, model 1/1
#> ✓ Fold01: recipe  6/10, model 1/1
#> i Fold01: recipe  6/10, model 1/1 (predictions)
#> i Fold01: recipe  7/10
#> ✓ Fold01: recipe  7/10
#> i Fold01: recipe  7/10, model 1/1
#> ✓ Fold01: recipe  7/10, model 1/1
#> i Fold01: recipe  7/10, model 1/1 (predictions)
#> i Fold01: recipe  8/10
#> ✓ Fold01: recipe  8/10
#> i Fold01: recipe  8/10, model 1/1
#> ✓ Fold01: recipe  8/10, model 1/1
#> i Fold01: recipe  8/10, model 1/1 (predictions)
#> i Fold01: recipe  9/10
#> ✓ Fold01: recipe  9/10
#> i Fold01: recipe  9/10, model 1/1
#> ✓ Fold01: recipe  9/10, model 1/1
#> i Fold01: recipe  9/10, model 1/1 (predictions)
#> i Fold01: recipe 10/10
#> ✓ Fold01: recipe 10/10
#> i Fold01: recipe 10/10, model 1/1
#> ✓ Fold01: recipe 10/10, model 1/1
#> i Fold01: recipe 10/10, model 1/1 (predictions)
#> i Fold02: recipe  1/10
#> ✓ Fold02: recipe  1/10
#> i Fold02: recipe  1/10, model 1/1
#> ✓ Fold02: recipe  1/10, model 1/1
#> i Fold02: recipe  1/10, model 1/1 (predictions)
#> i Fold02: recipe  2/10
#> ✓ Fold02: recipe  2/10
#> i Fold02: recipe  2/10, model 1/1
#> ✓ Fold02: recipe  2/10, model 1/1
#> i Fold02: recipe  2/10, model 1/1 (predictions)
#> i Fold02: recipe  3/10
#> ✓ Fold02: recipe  3/10
#> i Fold02: recipe  3/10, model 1/1
#> ✓ Fold02: recipe  3/10, model 1/1
#> i Fold02: recipe  3/10, model 1/1 (predictions)
#> i Fold02: recipe  4/10
#> ✓ Fold02: recipe  4/10
#> i Fold02: recipe  4/10, model 1/1
#> ✓ Fold02: recipe  4/10, model 1/1
#> i Fold02: recipe  4/10, model 1/1 (predictions)
#> i Fold02: recipe  5/10
#> ✓ Fold02: recipe  5/10
#> i Fold02: recipe  5/10, model 1/1
#> ✓ Fold02: recipe  5/10, model 1/1
#> i Fold02: recipe  5/10, model 1/1 (predictions)
#> i Fold02: recipe  6/10
#> ✓ Fold02: recipe  6/10
#> i Fold02: recipe  6/10, model 1/1
#> ✓ Fold02: recipe  6/10, model 1/1
#> i Fold02: recipe  6/10, model 1/1 (predictions)
#> i Fold02: recipe  7/10
#> ✓ Fold02: recipe  7/10
#> i Fold02: recipe  7/10, model 1/1
#> ✓ Fold02: recipe  7/10, model 1/1
#> i Fold02: recipe  7/10, model 1/1 (predictions)
#> i Fold02: recipe  8/10
#> ✓ Fold02: recipe  8/10
#> i Fold02: recipe  8/10, model 1/1
#> ✓ Fold02: recipe  8/10, model 1/1
#> i Fold02: recipe  8/10, model 1/1 (predictions)
#> i Fold02: recipe  9/10
#> ✓ Fold02: recipe  9/10
#> i Fold02: recipe  9/10, model 1/1
#> ✓ Fold02: recipe  9/10, model 1/1
#> i Fold02: recipe  9/10, model 1/1 (predictions)
#> i Fold02: recipe 10/10
#> ✓ Fold02: recipe 10/10
#> i Fold02: recipe 10/10, model 1/1
#> ✓ Fold02: recipe 10/10, model 1/1
#> i Fold02: recipe 10/10, model 1/1 (predictions)
#> i Fold03: recipe  1/10
#> ✓ Fold03: recipe  1/10
#> i Fold03: recipe  1/10, model 1/1
#> ✓ Fold03: recipe  1/10, model 1/1
#> i Fold03: recipe  1/10, model 1/1 (predictions)
#> i Fold03: recipe  2/10
#> ✓ Fold03: recipe  2/10
#> i Fold03: recipe  2/10, model 1/1
#> ✓ Fold03: recipe  2/10, model 1/1
#> i Fold03: recipe  2/10, model 1/1 (predictions)
#> i Fold03: recipe  3/10
#> ✓ Fold03: recipe  3/10
#> i Fold03: recipe  3/10, model 1/1
#> ✓ Fold03: recipe  3/10, model 1/1
#> i Fold03: recipe  3/10, model 1/1 (predictions)
#> i Fold03: recipe  4/10
#> ✓ Fold03: recipe  4/10
#> i Fold03: recipe  4/10, model 1/1
#> ✓ Fold03: recipe  4/10, model 1/1
#> i Fold03: recipe  4/10, model 1/1 (predictions)
#> i Fold03: recipe  5/10
#> ✓ Fold03: recipe  5/10
#> i Fold03: recipe  5/10, model 1/1
#> ✓ Fold03: recipe  5/10, model 1/1
#> i Fold03: recipe  5/10, model 1/1 (predictions)
#> i Fold03: recipe  6/10
#> ✓ Fold03: recipe  6/10
#> i Fold03: recipe  6/10, model 1/1
#> ✓ Fold03: recipe  6/10, model 1/1
#> i Fold03: recipe  6/10, model 1/1 (predictions)
#> i Fold03: recipe  7/10
#> ✓ Fold03: recipe  7/10
#> i Fold03: recipe  7/10, model 1/1
#> ✓ Fold03: recipe  7/10, model 1/1
#> i Fold03: recipe  7/10, model 1/1 (predictions)
#> i Fold03: recipe  8/10
#> ✓ Fold03: recipe  8/10
#> i Fold03: recipe  8/10, model 1/1
#> ✓ Fold03: recipe  8/10, model 1/1
#> i Fold03: recipe  8/10, model 1/1 (predictions)
#> i Fold03: recipe  9/10
#> ✓ Fold03: recipe  9/10
#> i Fold03: recipe  9/10, model 1/1
#> ✓ Fold03: recipe  9/10, model 1/1
#> i Fold03: recipe  9/10, model 1/1 (predictions)
#> i Fold03: recipe 10/10
#> ✓ Fold03: recipe 10/10
#> i Fold03: recipe 10/10, model 1/1
#> ✓ Fold03: recipe 10/10, model 1/1
#> i Fold03: recipe 10/10, model 1/1 (predictions)
#> i Fold04: recipe  1/10
#> ✓ Fold04: recipe  1/10
#> i Fold04: recipe  1/10, model 1/1
#> ✓ Fold04: recipe  1/10, model 1/1
#> i Fold04: recipe  1/10, model 1/1 (predictions)
#> i Fold04: recipe  2/10
#> ✓ Fold04: recipe  2/10
#> i Fold04: recipe  2/10, model 1/1
#> ✓ Fold04: recipe  2/10, model 1/1
#> i Fold04: recipe  2/10, model 1/1 (predictions)
#> i Fold04: recipe  3/10
#> ✓ Fold04: recipe  3/10
#> i Fold04: recipe  3/10, model 1/1
#> ✓ Fold04: recipe  3/10, model 1/1
#> i Fold04: recipe  3/10, model 1/1 (predictions)
#> i Fold04: recipe  4/10
#> ✓ Fold04: recipe  4/10
#> i Fold04: recipe  4/10, model 1/1
#> ✓ Fold04: recipe  4/10, model 1/1
#> i Fold04: recipe  4/10, model 1/1 (predictions)
#> i Fold04: recipe  5/10
#> ✓ Fold04: recipe  5/10
#> i Fold04: recipe  5/10, model 1/1
#> ✓ Fold04: recipe  5/10, model 1/1
#> i Fold04: recipe  5/10, model 1/1 (predictions)
#> i Fold04: recipe  6/10
#> ✓ Fold04: recipe  6/10
#> i Fold04: recipe  6/10, model 1/1
#> ✓ Fold04: recipe  6/10, model 1/1
#> i Fold04: recipe  6/10, model 1/1 (predictions)
#> i Fold04: recipe  7/10
#> ✓ Fold04: recipe  7/10
#> i Fold04: recipe  7/10, model 1/1
#> ✓ Fold04: recipe  7/10, model 1/1
#> i Fold04: recipe  7/10, model 1/1 (predictions)
#> i Fold04: recipe  8/10
#> ✓ Fold04: recipe  8/10
#> i Fold04: recipe  8/10, model 1/1
#> ✓ Fold04: recipe  8/10, model 1/1
#> i Fold04: recipe  8/10, model 1/1 (predictions)
#> i Fold04: recipe  9/10
#> ✓ Fold04: recipe  9/10
#> i Fold04: recipe  9/10, model 1/1
#> ✓ Fold04: recipe  9/10, model 1/1
#> i Fold04: recipe  9/10, model 1/1 (predictions)
#> i Fold04: recipe 10/10
#> ✓ Fold04: recipe 10/10
#> i Fold04: recipe 10/10, model 1/1
#> ✓ Fold04: recipe 10/10, model 1/1
#> i Fold04: recipe 10/10, model 1/1 (predictions)
#> i Fold05: recipe  1/10
#> ✓ Fold05: recipe  1/10
#> i Fold05: recipe  1/10, model 1/1
#> ✓ Fold05: recipe  1/10, model 1/1
#> i Fold05: recipe  1/10, model 1/1 (predictions)
#> i Fold05: recipe  2/10
#> ✓ Fold05: recipe  2/10
#> i Fold05: recipe  2/10, model 1/1
#> ✓ Fold05: recipe  2/10, model 1/1
#> i Fold05: recipe  2/10, model 1/1 (predictions)
#> i Fold05: recipe  3/10
#> ✓ Fold05: recipe  3/10
#> i Fold05: recipe  3/10, model 1/1
#> ✓ Fold05: recipe  3/10, model 1/1
#> i Fold05: recipe  3/10, model 1/1 (predictions)
#> i Fold05: recipe  4/10
#> ✓ Fold05: recipe  4/10
#> i Fold05: recipe  4/10, model 1/1
#> ✓ Fold05: recipe  4/10, model 1/1
#> i Fold05: recipe  4/10, model 1/1 (predictions)
#> i Fold05: recipe  5/10
#> ✓ Fold05: recipe  5/10
#> i Fold05: recipe  5/10, model 1/1
#> ✓ Fold05: recipe  5/10, model 1/1
#> i Fold05: recipe  5/10, model 1/1 (predictions)
#> i Fold05: recipe  6/10
#> ✓ Fold05: recipe  6/10
#> i Fold05: recipe  6/10, model 1/1
#> ✓ Fold05: recipe  6/10, model 1/1
#> i Fold05: recipe  6/10, model 1/1 (predictions)
#> i Fold05: recipe  7/10
#> ✓ Fold05: recipe  7/10
#> i Fold05: recipe  7/10, model 1/1
#> ✓ Fold05: recipe  7/10, model 1/1
#> i Fold05: recipe  7/10, model 1/1 (predictions)
#> i Fold05: recipe  8/10
#> ✓ Fold05: recipe  8/10
#> i Fold05: recipe  8/10, model 1/1
#> ✓ Fold05: recipe  8/10, model 1/1
#> i Fold05: recipe  8/10, model 1/1 (predictions)
#> i Fold05: recipe  9/10
#> ✓ Fold05: recipe  9/10
#> i Fold05: recipe  9/10, model 1/1
#> ✓ Fold05: recipe  9/10, model 1/1
#> i Fold05: recipe  9/10, model 1/1 (predictions)
#> i Fold05: recipe 10/10
#> ✓ Fold05: recipe 10/10
#> i Fold05: recipe 10/10, model 1/1
#> ✓ Fold05: recipe 10/10, model 1/1
#> i Fold05: recipe 10/10, model 1/1 (predictions)
#> i Fold06: recipe  1/10
#> ✓ Fold06: recipe  1/10
#> i Fold06: recipe  1/10, model 1/1
#> ✓ Fold06: recipe  1/10, model 1/1
#> i Fold06: recipe  1/10, model 1/1 (predictions)
#> i Fold06: recipe  2/10
#> ✓ Fold06: recipe  2/10
#> i Fold06: recipe  2/10, model 1/1
#> ✓ Fold06: recipe  2/10, model 1/1
#> i Fold06: recipe  2/10, model 1/1 (predictions)
#> i Fold06: recipe  3/10
#> ✓ Fold06: recipe  3/10
#> i Fold06: recipe  3/10, model 1/1
#> ✓ Fold06: recipe  3/10, model 1/1
#> i Fold06: recipe  3/10, model 1/1 (predictions)
#> i Fold06: recipe  4/10
#> ✓ Fold06: recipe  4/10
#> i Fold06: recipe  4/10, model 1/1
#> ✓ Fold06: recipe  4/10, model 1/1
#> i Fold06: recipe  4/10, model 1/1 (predictions)
#> i Fold06: recipe  5/10
#> ✓ Fold06: recipe  5/10
#> i Fold06: recipe  5/10, model 1/1
#> ✓ Fold06: recipe  5/10, model 1/1
#> i Fold06: recipe  5/10, model 1/1 (predictions)
#> i Fold06: recipe  6/10
#> ✓ Fold06: recipe  6/10
#> i Fold06: recipe  6/10, model 1/1
#> ✓ Fold06: recipe  6/10, model 1/1
#> i Fold06: recipe  6/10, model 1/1 (predictions)
#> i Fold06: recipe  7/10
#> ✓ Fold06: recipe  7/10
#> i Fold06: recipe  7/10, model 1/1
#> ✓ Fold06: recipe  7/10, model 1/1
#> i Fold06: recipe  7/10, model 1/1 (predictions)
#> i Fold06: recipe  8/10
#> ✓ Fold06: recipe  8/10
#> i Fold06: recipe  8/10, model 1/1
#> ✓ Fold06: recipe  8/10, model 1/1
#> i Fold06: recipe  8/10, model 1/1 (predictions)
#> i Fold06: recipe  9/10
#> ✓ Fold06: recipe  9/10
#> i Fold06: recipe  9/10, model 1/1
#> ✓ Fold06: recipe  9/10, model 1/1
#> i Fold06: recipe  9/10, model 1/1 (predictions)
#> i Fold06: recipe 10/10
#> ✓ Fold06: recipe 10/10
#> i Fold06: recipe 10/10, model 1/1
#> ✓ Fold06: recipe 10/10, model 1/1
#> i Fold06: recipe 10/10, model 1/1 (predictions)
#> i Fold07: recipe  1/10
#> ✓ Fold07: recipe  1/10
#> i Fold07: recipe  1/10, model 1/1
#> ✓ Fold07: recipe  1/10, model 1/1
#> i Fold07: recipe  1/10, model 1/1 (predictions)
#> i Fold07: recipe  2/10
#> ✓ Fold07: recipe  2/10
#> i Fold07: recipe  2/10, model 1/1
#> ✓ Fold07: recipe  2/10, model 1/1
#> i Fold07: recipe  2/10, model 1/1 (predictions)
#> i Fold07: recipe  3/10
#> ✓ Fold07: recipe  3/10
#> i Fold07: recipe  3/10, model 1/1
#> ✓ Fold07: recipe  3/10, model 1/1
#> i Fold07: recipe  3/10, model 1/1 (predictions)
#> i Fold07: recipe  4/10
#> ✓ Fold07: recipe  4/10
#> i Fold07: recipe  4/10, model 1/1
#> ✓ Fold07: recipe  4/10, model 1/1
#> i Fold07: recipe  4/10, model 1/1 (predictions)
#> i Fold07: recipe  5/10
#> ✓ Fold07: recipe  5/10
#> i Fold07: recipe  5/10, model 1/1
#> ✓ Fold07: recipe  5/10, model 1/1
#> i Fold07: recipe  5/10, model 1/1 (predictions)
#> i Fold07: recipe  6/10
#> ✓ Fold07: recipe  6/10
#> i Fold07: recipe  6/10, model 1/1
#> ✓ Fold07: recipe  6/10, model 1/1
#> i Fold07: recipe  6/10, model 1/1 (predictions)
#> i Fold07: recipe  7/10
#> ✓ Fold07: recipe  7/10
#> i Fold07: recipe  7/10, model 1/1
#> ✓ Fold07: recipe  7/10, model 1/1
#> i Fold07: recipe  7/10, model 1/1 (predictions)
#> i Fold07: recipe  8/10
#> ✓ Fold07: recipe  8/10
#> i Fold07: recipe  8/10, model 1/1
#> ✓ Fold07: recipe  8/10, model 1/1
#> i Fold07: recipe  8/10, model 1/1 (predictions)
#> i Fold07: recipe  9/10
#> ✓ Fold07: recipe  9/10
#> i Fold07: recipe  9/10, model 1/1
#> ✓ Fold07: recipe  9/10, model 1/1
#> i Fold07: recipe  9/10, model 1/1 (predictions)
#> i Fold07: recipe 10/10
#> ✓ Fold07: recipe 10/10
#> i Fold07: recipe 10/10, model 1/1
#> ✓ Fold07: recipe 10/10, model 1/1
#> i Fold07: recipe 10/10, model 1/1 (predictions)
#> i Fold08: recipe  1/10
#> ✓ Fold08: recipe  1/10
#> i Fold08: recipe  1/10, model 1/1
#> ✓ Fold08: recipe  1/10, model 1/1
#> i Fold08: recipe  1/10, model 1/1 (predictions)
#> i Fold08: recipe  2/10
#> ✓ Fold08: recipe  2/10
#> i Fold08: recipe  2/10, model 1/1
#> ✓ Fold08: recipe  2/10, model 1/1
#> i Fold08: recipe  2/10, model 1/1 (predictions)
#> i Fold08: recipe  3/10
#> ✓ Fold08: recipe  3/10
#> i Fold08: recipe  3/10, model 1/1
#> ✓ Fold08: recipe  3/10, model 1/1
#> i Fold08: recipe  3/10, model 1/1 (predictions)
#> i Fold08: recipe  4/10
#> ✓ Fold08: recipe  4/10
#> i Fold08: recipe  4/10, model 1/1
#> ✓ Fold08: recipe  4/10, model 1/1
#> i Fold08: recipe  4/10, model 1/1 (predictions)
#> i Fold08: recipe  5/10
#> ✓ Fold08: recipe  5/10
#> i Fold08: recipe  5/10, model 1/1
#> ✓ Fold08: recipe  5/10, model 1/1
#> i Fold08: recipe  5/10, model 1/1 (predictions)
#> i Fold08: recipe  6/10
#> ✓ Fold08: recipe  6/10
#> i Fold08: recipe  6/10, model 1/1
#> ✓ Fold08: recipe  6/10, model 1/1
#> i Fold08: recipe  6/10, model 1/1 (predictions)
#> i Fold08: recipe  7/10
#> ✓ Fold08: recipe  7/10
#> i Fold08: recipe  7/10, model 1/1
#> ✓ Fold08: recipe  7/10, model 1/1
#> i Fold08: recipe  7/10, model 1/1 (predictions)
#> i Fold08: recipe  8/10
#> ✓ Fold08: recipe  8/10
#> i Fold08: recipe  8/10, model 1/1
#> ✓ Fold08: recipe  8/10, model 1/1
#> i Fold08: recipe  8/10, model 1/1 (predictions)
#> i Fold08: recipe  9/10
#> ✓ Fold08: recipe  9/10
#> i Fold08: recipe  9/10, model 1/1
#> ✓ Fold08: recipe  9/10, model 1/1
#> i Fold08: recipe  9/10, model 1/1 (predictions)
#> i Fold08: recipe 10/10
#> ✓ Fold08: recipe 10/10
#> i Fold08: recipe 10/10, model 1/1
#> ✓ Fold08: recipe 10/10, model 1/1
#> i Fold08: recipe 10/10, model 1/1 (predictions)
#> i Fold09: recipe  1/10
#> ✓ Fold09: recipe  1/10
#> i Fold09: recipe  1/10, model 1/1
#> ✓ Fold09: recipe  1/10, model 1/1
#> i Fold09: recipe  1/10, model 1/1 (predictions)
#> i Fold09: recipe  2/10
#> ✓ Fold09: recipe  2/10
#> i Fold09: recipe  2/10, model 1/1
#> ✓ Fold09: recipe  2/10, model 1/1
#> i Fold09: recipe  2/10, model 1/1 (predictions)
#> i Fold09: recipe  3/10
#> ✓ Fold09: recipe  3/10
#> i Fold09: recipe  3/10, model 1/1
#> ✓ Fold09: recipe  3/10, model 1/1
#> i Fold09: recipe  3/10, model 1/1 (predictions)
#> i Fold09: recipe  4/10
#> ✓ Fold09: recipe  4/10
#> i Fold09: recipe  4/10, model 1/1
#> ✓ Fold09: recipe  4/10, model 1/1
#> i Fold09: recipe  4/10, model 1/1 (predictions)
#> i Fold09: recipe  5/10
#> ✓ Fold09: recipe  5/10
#> i Fold09: recipe  5/10, model 1/1
#> ✓ Fold09: recipe  5/10, model 1/1
#> i Fold09: recipe  5/10, model 1/1 (predictions)
#> i Fold09: recipe  6/10
#> ✓ Fold09: recipe  6/10
#> i Fold09: recipe  6/10, model 1/1
#> ✓ Fold09: recipe  6/10, model 1/1
#> i Fold09: recipe  6/10, model 1/1 (predictions)
#> i Fold09: recipe  7/10
#> ✓ Fold09: recipe  7/10
#> i Fold09: recipe  7/10, model 1/1
#> ✓ Fold09: recipe  7/10, model 1/1
#> i Fold09: recipe  7/10, model 1/1 (predictions)
#> i Fold09: recipe  8/10
#> ✓ Fold09: recipe  8/10
#> i Fold09: recipe  8/10, model 1/1
#> ✓ Fold09: recipe  8/10, model 1/1
#> i Fold09: recipe  8/10, model 1/1 (predictions)
#> i Fold09: recipe  9/10
#> ✓ Fold09: recipe  9/10
#> i Fold09: recipe  9/10, model 1/1
#> ✓ Fold09: recipe  9/10, model 1/1
#> i Fold09: recipe  9/10, model 1/1 (predictions)
#> i Fold09: recipe 10/10
#> ✓ Fold09: recipe 10/10
#> i Fold09: recipe 10/10, model 1/1
#> ✓ Fold09: recipe 10/10, model 1/1
#> i Fold09: recipe 10/10, model 1/1 (predictions)
#> i Fold10: recipe  1/10
#> ✓ Fold10: recipe  1/10
#> i Fold10: recipe  1/10, model 1/1
#> ✓ Fold10: recipe  1/10, model 1/1
#> i Fold10: recipe  1/10, model 1/1 (predictions)
#> i Fold10: recipe  2/10
#> ✓ Fold10: recipe  2/10
#> i Fold10: recipe  2/10, model 1/1
#> ✓ Fold10: recipe  2/10, model 1/1
#> i Fold10: recipe  2/10, model 1/1 (predictions)
#> i Fold10: recipe  3/10
#> ✓ Fold10: recipe  3/10
#> i Fold10: recipe  3/10, model 1/1
#> ✓ Fold10: recipe  3/10, model 1/1
#> i Fold10: recipe  3/10, model 1/1 (predictions)
#> i Fold10: recipe  4/10
#> ✓ Fold10: recipe  4/10
#> i Fold10: recipe  4/10, model 1/1
#> ✓ Fold10: recipe  4/10, model 1/1
#> i Fold10: recipe  4/10, model 1/1 (predictions)
#> i Fold10: recipe  5/10
#> ✓ Fold10: recipe  5/10
#> i Fold10: recipe  5/10, model 1/1
#> ✓ Fold10: recipe  5/10, model 1/1
#> i Fold10: recipe  5/10, model 1/1 (predictions)
#> i Fold10: recipe  6/10
#> ✓ Fold10: recipe  6/10
#> i Fold10: recipe  6/10, model 1/1
#> ✓ Fold10: recipe  6/10, model 1/1
#> i Fold10: recipe  6/10, model 1/1 (predictions)
#> i Fold10: recipe  7/10
#> ✓ Fold10: recipe  7/10
#> i Fold10: recipe  7/10, model 1/1
#> ✓ Fold10: recipe  7/10, model 1/1
#> i Fold10: recipe  7/10, model 1/1 (predictions)
#> i Fold10: recipe  8/10
#> ✓ Fold10: recipe  8/10
#> i Fold10: recipe  8/10, model 1/1
#> ✓ Fold10: recipe  8/10, model 1/1
#> i Fold10: recipe  8/10, model 1/1 (predictions)
#> i Fold10: recipe  9/10
#> ✓ Fold10: recipe  9/10
#> i Fold10: recipe  9/10, model 1/1
#> ✓ Fold10: recipe  9/10, model 1/1
#> i Fold10: recipe  9/10, model 1/1 (predictions)
#> i Fold10: recipe 10/10
#> ✓ Fold10: recipe 10/10
#> i Fold10: recipe 10/10, model 1/1
#> ✓ Fold10: recipe 10/10, model 1/1
#> i Fold10: recipe 10/10, model 1/1 (predictions)

show_best(res, metric = "rmse")
#> # A tibble: 5 x 11
#>   `lon breaks` `lon depth` `lon min_n` `lat breaks` `lat depth` `lat min_n`
#>          <int>       <int>       <int>        <int>       <int>       <int>
#> 1            7          10          14            8          10          17
#> 2           14           8           4            6          14           3
#> 3           28           6           8           27          12           8
#> 4           22          14          28           15           8          13
#> 5            9           4          11           28           4          20
#> # … with 5 more variables: .metric <chr>, .estimator <chr>, mean <dbl>,
#> #   n <int>, std_err <dbl>
```

<sup>Created on 2020-05-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

konradsemsch commented:
Thanks for answering the questions regarding `tunable`, it's clear to me now.

Ok, I will work on prop and will also do a little bit more testing with tuning.

Note on the `learn_rate` - it's actually not really a parameter that should be tuned. I see value in exposing it to the user to give some sense of control over the function, but in practice all other parameters could be tuned while keeping `learn_rate` constant. My view is that we probably should restrict it between 0.05 and 0.3, as it should be the function of how long learning takes with regards to the size of the data.

So in practice, the user first would keep `learn_rate` high (0.3) to tune other parameters, and while their optimal values are found, once could finalize the workflow with a lower `learn_rate` of e.g. 0.1.

topepo commented:
Can you give me permissions to push to your repo directly?

konradsemsch commented:
You should receive an invite

konradsemsch commented:
I used your testing code and modified it to run with a very small learning rate (0.001) which resulted in the following error:

```r
"recipe  1/10: `step_discretize_xgb()` failed to create a tree with error for predictor 'Longitude', which will not be binned. The error: Non-tree model detected! This function can only be used with tree models., `step_discretize_xgb()` failed to create a tree with error for predictor 'Latitude', which will not be binned. The error: Non-tree model detected! This function can only be used with tree models."
```

If you take a look at the main code, I already tackled this issue before:

```r
  # Changes: if there is insufficient training data/ variation then xgboost model is constant
  # and no splits will be returned. Additional check will inform the user
  # that the dataset is insufficient for this particular case
  # https://github.com/dmlc/xgboost/issues/2876
  # https://stackoverflow.com/questions/42670033/r-getting-non-tree-model-detected-this-function-can-only-be-used-with-tree-mo
  xgb_tree <- try(
    xgboost::xgb.model.dt.tree(
      model = xgb_mdl,
      trees = 1:xgb_mdl$best_iteration,
      use_int_id = TRUE
    ),
    silent = TRUE
  )
```
Increasing the `learn_rate` from 0.001 to 0.005 solves the issue... So I think the answer to this problem is quite simple: we should provide a warning message to the user to slightly increase the learning rate, as it is too small compared to the size of the data in order for the algo to learn sufficiently in subsequent steps.

```r
library(tidymodels)
library(AmesHousing)
library(embed)

ames <-
  make_ames() %>%
  # Remove quality-related predictors
  dplyr::select(-matches("Qu"))

set.seed(4595)
data_split <- initial_split(ames, strata = "Sale_Price")
ames_train <- training(data_split)
ames_test  <- testing(data_split)
rs <- vfold_cv(ames_train)

ames_rec <-
  recipe(Sale_Price ~ Longitude + Latitude,
         data = ames_train) %>%
  step_log(Sale_Price, base = 10) %>%
  step_discretize_xgb(
    all_numeric(),
    outcome = "Sale_Price",
    num_breaks = tune(),
    tree_depth = tune(),
    min_n = tune(),
    learn_rate = 0.001 ### just slightly increase it to e.g. 0.005
  ) %>%
  step_zv(all_predictors())

mod <- linear_reg() %>% set_engine("lm")

set.seed(8199)
res <-
  mod %>% tune_grid(
    ames_rec,
    resamples = rs,
    grid = 10,
    control = control_grid(verbose = TRUE)
  )
```

konradsemsch commented:
Thanks for the updates! I have one question though: why did you put the learning rate on a log10 scale?

topepo commented:
> why did you put the learning rate on a log10 scale?

It is a parameter that is (in general) greater than zero and can vary across different orders of magnitude. Our `dials` function uses a log-10 transform when simulating or making a grid. For that reason, the `range` declaration in `tunable()` should be in log-10 units. 

Are you generally happy with where this are at this point? I'm good except for some testing issues. 

The [test failures](https://github.com/tidymodels/embed/actions/runs/113434907) have consistently been about the specific bins that we check for. I think that we'll have to do checks that are less literal than the specific break points. They are also different in R >3.5 and R <= 3.5 due to the change in the base R rng.

konradsemsch commented:
Ok, thanks for the explanation.

I did a review of both xgb and cart methods and corrected some minor things/ added notes. I think our implementation makes sense as of now.

However, there's just one thing I've been kicking around, namely:  if it makes sense to expose number of training steps and early stopping rounds in xgboost. I think for a lot of problems this amount of steps is more than enough (100 per single variable), but I can imagine some users may want to play with them as well in some situations. If we're exposing the learning rate, exposing these 2 parameters would make sense too (not for tuning though). Additionally, it may make sense to inform the user how many steps it took for each variable to find optimal splits. What do you think?

I also did some testing for both methods using all numeric variables to see if it's behaving ok. When using all variables as is, the output is very rich on the notes, but I think generally correct. You can take a look below.

**xgb**

```r
library(tidymodels)
library(AmesHousing)
library(embed)

ames <-
  make_ames() %>%
  # Remove quality-related predictors
  dplyr::select(-matches("Qu"))

set.seed(4595)
data_split <- initial_split(ames, strata = "Sale_Price")
ames_train <- training(data_split)
ames_test  <- testing(data_split)
rs <- vfold_cv(ames_train, v = 5)

ames_rec <-
  recipe(Sale_Price ~ ., data = ames_train) %>%
  step_log(Sale_Price, base = 10) %>%
  step_other(all_nominal(), threshold = 0.01) %>% 
  step_discretize_xgb(
    all_numeric(),
    outcome = "Sale_Price",
    num_breaks = tune(),
    tree_depth = tune(),
    min_n = tune()
  ) %>%
  step_zv(all_predictors())

mod <- linear_reg() %>% set_engine("lm")

set.seed(8199)
res <-
  mod %>% tune_grid(
    ames_rec,
    resamples = rs,
    grid = 5,
    control = control_grid(verbose = TRUE)
  )

show_best(res, "rmse", 1)
# A tibble: 1 x 8
  num_breaks tree_depth min_n .metric .estimator   mean     n std_err
       <int>      <int> <int> <chr>   <chr>       <dbl> <int>   <dbl>
1         13          4    15 rmse    standard   0.0610     5 0.00186
```

```r
> res$.notes[[1]][[1]]
 [1] "recipe 1/5: More than 20 unique training set values are required. Predictors 'BsmtFin_SF_1', 'Bsmt_Full_Bath', 'Bsmt_Half_Bath', 'Full_Bath', 'Half_Bath', 'Bedroom_AbvGr', 'Kitchen_AbvGr', 'TotRms_AbvGrd', 'Fireplaces', 'Garage_Cars', 'Pool_Area', 'Mo_Sold', 'Year_Sold' were not processed; their original values will be used., `step_discretize_xgb()` failed for predictor 'Three_season_porch'. This could be because the data have no trend or because the learning rate is too low (current value: 0.3). The predictor was not binned."
 [2] "recipe 1/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                        
 [3] "recipe 2/5: More than 20 unique training set values are required. Predictors 'BsmtFin_SF_1', 'Bsmt_Full_Bath', 'Bsmt_Half_Bath', 'Full_Bath', 'Half_Bath', 'Bedroom_AbvGr', 'Kitchen_AbvGr', 'TotRms_AbvGrd', 'Fireplaces', 'Garage_Cars', 'Pool_Area', 'Mo_Sold', 'Year_Sold' were not processed; their original values will be used."                                                                                                                                                                                                             
 [4] "recipe 2/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                        
 [5] "recipe 3/5: More than 20 unique training set values are required. Predictors 'BsmtFin_SF_1', 'Bsmt_Full_Bath', 'Bsmt_Half_Bath', 'Full_Bath', 'Half_Bath', 'Bedroom_AbvGr', 'Kitchen_AbvGr', 'TotRms_AbvGrd', 'Fireplaces', 'Garage_Cars', 'Pool_Area', 'Mo_Sold', 'Year_Sold' were not processed; their original values will be used., `step_discretize_xgb()` failed for predictor 'Three_season_porch'. This could be because the data have no trend or because the learning rate is too low (current value: 0.3). The predictor was not binned."
 [6] "recipe 3/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                        
 [7] "recipe 4/5: More than 20 unique training set values are required. Predictors 'BsmtFin_SF_1', 'Bsmt_Full_Bath', 'Bsmt_Half_Bath', 'Full_Bath', 'Half_Bath', 'Bedroom_AbvGr', 'Kitchen_AbvGr', 'TotRms_AbvGrd', 'Fireplaces', 'Garage_Cars', 'Pool_Area', 'Mo_Sold', 'Year_Sold' were not processed; their original values will be used."                                                                                                                                                                                                             
 [8] "recipe 4/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                        
 [9] "recipe 5/5: More than 20 unique training set values are required. Predictors 'BsmtFin_SF_1', 'Bsmt_Full_Bath', 'Bsmt_Half_Bath', 'Full_Bath', 'Half_Bath', 'Bedroom_AbvGr', 'Kitchen_AbvGr', 'TotRms_AbvGrd', 'Fireplaces', 'Garage_Cars', 'Pool_Area', 'Mo_Sold', 'Year_Sold' were not processed; their original values will be used."                                                                                                                                                                                                             
[10] "recipe 5/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading" 
```

**cart**

```r
ames_rec <-
  recipe(Sale_Price ~ ., data = ames_train) %>%
  step_log(Sale_Price, base = 10) %>%
  step_other(all_nominal(), threshold = 0.01) %>% 
  step_discretize_cart(
    all_numeric(),
    outcome = "Sale_Price",
    cost_complexity = tune(),
    tree_depth = tune(),
    min_n = tune()
  ) %>%
  step_zv(all_predictors())

mod <- linear_reg() %>% set_engine("lm")

set.seed(8199)
res <-
  mod %>% tune_grid(
    ames_rec,
    resamples = rs,
    grid = 5,
    control = control_grid(verbose = TRUE)
  )

show_best(res, "rmse", 1)
# A tibble: 1 x 8
  cost_complexity tree_depth min_n .metric .estimator   mean     n std_err
            <dbl>      <int> <int> <chr>   <chr>       <dbl> <int>   <dbl>
1         0.00899          4    23 rmse    standard   0.0618     5 0.00162
```

```r
> res$.notes[[1]][[1]]
[1] "recipe 1/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
[2] "recipe 2/5: `step_discretize_cart()` failed to find any meaningful splits for predictor 'BsmtFin_SF_2', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Bsmt_Half_Bath', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Three_season_porch', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Pool_Area', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Misc_Val', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Mo_Sold', which will not be binned., `step_discretize_cart()` failed to find any meaningful splits for predictor 'Year_Sold', which will not be binned."
[3] "recipe 2/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
[4] "recipe 3/5: `step_discretize_cart()` failed to find any meaningful splits for predictor 'Pool_Area', which will not be binned."                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
[5] "recipe 3/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
[6] "recipe 4/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
[7] "recipe 5/5, model 1/1 (predictions): prediction from a rank-deficient fit may be misleading" 
```

topepo commented:
A lot of times when we aren't sure about including things we release it and see if anyone adds an issue. Your call though. 

Those particular predictors that failed were unlikely to produce splits. 

I'm read to merge if you are.

konradsemsch commented:
> Those particular predictors that failed were unlikely to produce splits.

Yes, that was my thought as well.

> A lot of times when we aren't sure about including things we release it and see if anyone adds an issue. Your call though.

Let's follow this approach and merge. Thanks for your help on this :)

topepo commented:
I appreciate your help!
